### ISSUE TITLE
Non-existent value in website-agent

### ISSUE THREAD

**Issue by robbielj** [cantino/huginn#856]
Sometimes when the agent cannot extract a value, it will mark it as `[]`, which prompts uneven matches and stops event generating. Is it a good practice to convert it into an empty value like `['']`? In this way, events can be created as usual. 

For example, change the jsonpath extract:
`result = Utils.values_at(doc, extraction_details['path'])`
to
`result = (Utils.values_at(doc, extraction_details['path']) == [] ? [''] : Utils.values_at(doc, extraction_details['path']))`

well, i am not sure whether it is a design on purpose.

**Comment by cantino** [comment]
Hey @robbielj!

Is this a case where you're extracting multiple groups, but only some of them have a certain field?

Could you provide an example?

**Comment by robbielj** [comment]
hi, @cantino 

for example, the captured JSON response is like

```
{
    "stuff": {
        "key1": val1,
        "key2": val2,
        "key3": val3,
        "key4": val4,
        "key5": val5,
     }
}
```
I want to extract all these values to create events. so the paths are `stuff.key1`, `stuff.key2` etc. Under specific circumstance, one of the keys may be missing. Currently the agent will report uneven matches and stops creating events. Is it better that the value of the missing key is set to `['']` (empty) as default so that event rolling-out can continue? 

This case may also apply to other scraping sources, when one of extracted field is missing.

**Comment by cantino** [comment]
Could you make your selector be an ORed list of the possible keys?  Is this XPath or JSONPath?

**Comment by robbielj** [comment]
I tried JSONPath but cannot get a clue on how to do an "OR" between an possibly non-existent path and a string like "". Any suggestion? 

I have switched `ALLOW_JSONPATH_EVAL` on.

**Comment by cantino** [comment]
I assume `$.stuff.*` wouldn't work because there are some keys you don't want?

**Comment by robbielj** [comment]
Indeed, I just want a few keys of the response.

**Comment by cantino** [comment]
@dsander Do you know if it's possible to specify a list of keys in JSONPath?

**Comment by dsander** [comment]
@cantino I dont think that is possible.

@robbielj Can you use a `EventFormattingAgent` after fetching all keys to select just the ones you are interested in?

**Comment by robbielj** [comment]
Thanks @dsander. `EventFormattingAgent` is a good idea!  Just checked and it works.

This issue can be closed now.

**Comment by robbielj** [comment]
Hi

i re-open this issue as i feel adding an extra event formatting agent for each scraping instance is quite inefficient. I suppose data fields could be optional thus should not interrupt event creating (uneven number of matches error).

Another possible approach is to set a default value for cases when no value is found following scraping rules.

**Comment by cantino** [comment]
The challenge with using a default value for a matcher is that you don't know which groups any present values belong to.  For example, say you want to match

```
title: hi
text: 123

title: hello

title: hi again
text: 234
```

if you have a `title` and `text` matchers, how does it know how to group the three matches of title and the two of text?

**Comment by robbielj** [comment]
I am not sure whether i follow your idea. The below is my thought. 
The matches are stored into an array (list) for each data field. for example:

normal case. there is no missing "title" and "url".

```
title
["title1", "title2", "title3", "title4", "title5"]
url
["url1", "url2", “url3", "url4", "url5"]
```

This case will report uneven matches error since "title3" is missing. array lengths are not consistent across data fields.
```
title
["title1", "title2", "title4", "title5"]
url
["url1", "url2", "url3", "url4", "url5"]
```

if we set a default value for "title". here set as ""
```
title
["title1", "title2", "", "title4", "title5"]
url
["url1", "url2", "url3", "url4", "url5"]

```
this should prevent the uneven matches error.

**Comment by cantino** [comment]
Right, the problem is that we don't know which url that missing title should be associated with.  What we have is two arrays, one that has 4 items and one with 5.  How do we know where in the 4 item array we should insert a blank default item to make them the same length?

**Comment by K1773R** [comment]
if you parse a list of things, its usually something nested in a div. we could set the upper div (a div that repeats for each item) to limit the searching. ie, if its not in this div with class/id foo, then the value should be empty.
but i that would need something new, i doubt xpath can do that.

**Comment by cantino** [comment]
Yea, if we allowed nesting of matcher groups, this would be simpler.  That might end up looking a lot like https://github.com/fizx/parsley.

**Comment by gdomod** [comment]
any solution for that ? maybe default value ?

**Comment by K1773R** [comment]
any progress on this?

**Comment by cantino** [comment]
I'm open to PRs from anyone who'd like to tackle this.  If you want to discuss possible solutions, feel free to comment here.

**Comment by hunhejj** [comment]
Hi all,
I think I am looking for something similar, but maybe it can be simply solved with using an appropriate jsonPath.
Consider the following example (json):
```
{
    "books": [
        {"title": "Title1", "author": {"name": "foo"}},
        {"title": "Title2", "author": {"name": "bar"}},
        {"title": "Title3"},
    ]
}
```
I want to extract the title of each book and the author's name (or an empty string if not present) using a WebsiteAgent.
Currently I would do this:
```
"extract": {
  "title": { "path": "$.books[*].title" },
  "name": { "path": "$.books[*].author.name" }
}
```
Given the above example this would result in an error (uneven number of matches, since the third book has no author).
Any ideas how this could be solved? Is there a way to specify an OR condition with jsonPath, e.g. `path1 or "" if path 1 does not exist`?

**Comment by cantino** [comment]
Hey @hunhejj, that's a good idea. I don't think you can do an OR currently in JSONPath, though.

The problem is similar to the case with HTML. If we get 4 titles and 5 names, which of the names should be missing a title?  We have to add a concept of a container or parent that should always match N times, and then it becomes clear which ones don't contain a certain child.

**Comment by hunhejj** [comment]
Oh I see. I will check if I can find any other solution. Which version of JSONPath does the project use?

**Comment by cantino** [comment]
We use a fork maintained by @Skarlso. Recently updated in #1674. https://github.com/Skarlso/jsonpathv2

**Comment by Skarlso** [comment]
I hear the summons. ![troll](http://classic.battle.net/war3/images/orc/units/portraits/headhunter.gif)

**Comment by blackberryoctopus** [comment]
@cantino @dsander 

I am also searching for a way to deal with a similar 'uneven number of matches' issue and hoping there may be a workaround. 

The agent below is expected to find new github repositories containing the string 'jacquard'. What I've found though is the layout engine for github.com only generates a col-9 class <div> when the repo owner enters the 'about' information for their project. 

*Note the lack of this for field for  "project-renard/p5-Renard-Jacquard"


`{
  "expected_update_period_in_days": "2",
  "url": "https://github.com/search?o=desc&q=jacquard&s=updated&type=Repositories&utf8=%E2%9C%93",
  "type": "html",
  "mode": "on_change",
  "extract": {
    "repo_name": {
      "css": ".col-9",
      "value": "normalize-space(.)"
    },
    "repo_url": {
      "css": ".col-8 h3",
      "value": "normalize-space(.)"
    }
  }
}`

**Comment by Skarlso** [comment]
@blackberryoctopus now you Can use OR in your JSONpath query.

**Comment by blackberryoctopus** [comment]
@Skarlso Where can I find the syntax documentation for using 'OR' in the JSONpath query?

**Comment by Skarlso** [comment]
@blackberryoctopus https://github.com/joshbuddy/jsonpath#conditional-operators-are-also-supported

**Comment by blackberryoctopus** [comment]
@Skarlso  thanks for sharing the URL to the documentation. 

I have one more question: My example agent receives a HTML document; my understanding is that CSS selectors are appropriate for searching the HTML document type. 

Are you implying that JSONPath expressions are also a valid search mechanism for HTML documents?

**Comment by Skarlso** [comment]
Hey man.

No, unfortunately JSONpath only formats JSON. For html CSS would be a wise choice.

**Comment by blackberryoctopus** [comment]
Thanks for the info @skarlso; this brings us back to my original question. 

Is there a mechanism by which CSS selector syntax may be used to more flexibly handle scenarios where a html element may or may not appear in a given document, essentially a ternary operator that assigns a default value.

e.g.

I want to capture the following the following objects:
`
{
property1 = .ganymede h3 
property 2 = (.rhea)?.rhea:'' //ternary operator
}
`

Such that the below HTML would render the following objects generated:
`
{
property1 = 'textX'
property2 = 'textY'
},
{
property1 = 'textZ'
property2 = ''
}
`

From the following html: 

`
<div class="japetus">
  <div class="ganymede">
    <h3>
      <a href="/saturn" class="titan">textX></a>
    </h3>
      <p class="rhea">textY</p>
	</div>
</div>
<div class="japetus">
  <div class="ganymede">
    <h3>
      <a href="/saturn2" class="titan">textZ</em></a>
    </h3>
 </div>
</div>
`

**Comment by dsander** [comment]
@blackberryoctopus I don't think that this is possible with the `WebsiteAgent` at the moment. You could use the [GitHub search API](https://developer.github.com/v3/search/) and process the results with EventFormattingAgent downstream.

**Comment by blackberryoctopus** [comment]
@dsander  Thank you for clarifying. I've implemented the extraction of data with the API as you suggested and processed downstream in a formatting agent.

**Comment by Ralphbupt** [comment]
@cantino  as @hunhejj  has said in situation like this 
{
    "books": [
        {"title": "Title1", "author": {"name": "foo"}},
        {"title": "Title2", "author": {"name": "bar"}},
        {"title": "Title3"},
    ]
}
and with extract

"extract": {
  "title": { "path": "$.books[*].title" },
  "name": { "path": "$.books[*].author.name" }
}

in this situation, we got three "title" and two "author", you mean it's difficult to decide which "title" should miss the "author", but as we can see , the first and second "title" and "author" are in the same {}, or same level of  XML, is it possible to check the {} or the same level of XML to find whether they have the corresponding attribute

**Comment by bingyupiaoyao** [comment]
Is there any progress about that issue?

**Comment by yefoenix** [comment]
Is there any progress about that issue?

**Comment by xiaoshikun801** [comment]
Is there any progress about that issue?

**Comment by avrilmaomao** [comment]
It it possible to send an email notification when a website-agent failed to extract data? For situations where the url is broken or return unexpected results?

**Comment by tarek369** [comment]
any updates ?

**Comment by tarek369** [comment]
I tried 
    "price": {
      "xpath": ".//div[contains(@class,\"oi-bottom clear\")]//p[contains(@class,\"price\")]/span/text()[1]",
      "value": "normalize-space(.)",
      "array": "true"

and it helped

**Comment by klausmcm** [comment]
> Thanks for the info @Skarlso; this brings us back to my question.
> 
> Is there a mechanism by which CSS selector syntax may be used to more flexibly handle scenarios where a html element may or may not appear in a given document, essentially a ternary operator that assigns a default value.
> 
> e.g.
> 
> I want to capture the following the following content from a HTML document:
> `{ property1 = .ganymede h3 property2 = (.rhea)?.rhea:'' //ternary operator }`
> 
> And generate the following Huginn message objects:
> `{ property1 = 'textX' property2 = 'textY' }, { property1 = 'textZ' property2 = '' }`
> 
> Using this made-up example html:
> [snippet.txt](https://github.com/huginn/huginn/files/1525048/snippet.txt)

I'd also love to know if there is a recommended way to deal with this. I'm trying to get URLs and most of them have images associated with them. Only a few don't and this breaks my website agent at the moment.

**Comment by haiphamhoang** [comment]
I have the same issue, my website agent stops generating events when the values extracted: []
`[00:00:00] INFO -- : Values extracted: []`
`[00:00:00] INFO -- : Dry Run finished`

**Comment by fireshaper** [comment]
Also having a similar issue. Would be nice if there was a "default" key we could add to extract. Something like
```
"image": {
      "xpath": "/html/body/div/img/@src",
      "value": ".",
      "default": "https://www.google.com/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png"
    }
```

**Comment by aooaa** [comment]
Any progress?
I got the same problem when I scrape the website for a series value, the problem is the website sometimes don't have some of the key and value, if this happen, the website agent will be terminated. what if huginn can add a judgment, when miss the search key and value, user can define the return a fixed value as an option, so the website agent can still process the next item.

**Comment by virtadpt** [comment]
That's a really good idea.

**Comment by ReVoLt112** [comment]
> Any progress? I got the same problem when I scrape the website for a series value, the problem is the website sometimes don't have some of the key and value, if this happen, the website agent will be terminated. what if huginn can add a judgment, when miss the search key and value, user can define the return message with a fixed value as an option, so the website agent can still process the next item.

Please add this feature!
